File: csharp-aspnetcore-sample/Models/Listing.cs
using System.ComponentModel.DataAnnotations;
using System.Runtime.Serialization;

namespace csharp_aspnetcore_sample.Models;

public enum Category {
    [EnumMember(Value = "Electronics")]
    Electronics,
    [EnumMember(Value = "Fashion")]
    Fashion,
    [EnumMember(Value = "Home & Garden")]
    HomeAndGarden,
    [EnumMember(Value = "Motors")]
    Motors,
    [EnumMember(Value = "Collectibles & Art")]
    CollectiblesAndArt,
    [EnumMember(Value = "Sporting Goods")]
    SportingGoods,
    [EnumMember(Value = "Toys & Hobbies")]
    ToysAndHobbies,
    [EnumMember(Value = "Business & Industrial")]
    BusinessAndIndustrial,
    [EnumMember(Value = "Music")]
    Music,
    [EnumMember(Value = "Health & Beauty")]
    HealthAndBeauty,
    [EnumMember(Value = "Books")]
    Books,
    [EnumMember(Value = "Cameras & Photo")]
    CamerasAndPhoto,
    [EnumMember(Value = "Computers, Tablets & Networking")]
    ComputersTabletsAndNetworking,
    [EnumMember(Value = "Cell Phones & Accessories")]
    CellPhonesAndAccessories,
    [EnumMember(Value = "Video Games & Consoles")]
    VideoGamesAndConsoles
}

public class Listing {
    public Guid ListingId { get; set; }

    [Required]
    [StringLength(100)]
    public string Name { get; set; } = string.Empty;

    [Required]
    public string Description { get; set; } = string.Empty;

    [Required]
    public Price Price { get; set; } = new();

    [Required]
    public Category Category { get; set; }

    [Required]
    public Location Location { get; set; } = new();
}

public class Price {
    [Required]
    [StringLength(3, MinimumLength = 3)]
    public string Currency { get; set; } = string.Empty;

    [Required]
    [Range(0.01, double.MaxValue, ErrorMessage = "Amount must be greater than 0")]
    public decimal Amount { get; set; }
}

public class Location {
    [Required]
    [StringLength(2, MinimumLength = 2, ErrorMessage = "Country must be a valid ISO 3166 2-character code")]
    public string Country { get; set; } = string.Empty;

    [Required]
    public string Municipality { get; set; } = string.Empty;

    [Required]
    [StringLength(7, MinimumLength = 7, ErrorMessage = "Geohash must be exactly 7 characters")]
    public string Geohash { get; set; } = string.Empty;
}

public class PaginatedListingsResponse {
    public IEnumerable<Listing> Items { get; set; } = new List<Listing>();
    public int TotalItems { get; set; }
    public int Page { get; set; }
    public int PageSize { get; set; }
}

public class FilterRequest {
    public List<Filter> Filters { get; set; } = new();
    public int Page { get; set; } = 1;
    public int PageSize { get; set; } = 10;
}

public class Filter {
    [Required]
    public string Field { get; set; } = string.Empty;
    
    [Required]
    public string Operator { get; set; } = string.Empty;
    
    [Required]
    public object Value { get; set; } = string.Empty;
}

public class SearchResponse {
    public IEnumerable<Listing> Items { get; set; } = new List<Listing>();
    public int TotalItems { get; set; }
    public int Page { get; set; }
    public int PageSize { get; set; }
    public List<Filter> AppliedFilters { get; set; } = new();
}

File: kotlin-ktor-sample/src/main/kotlin/com/example/models/Listing.kt
package com.example.models

import java.math.BigDecimal
import java.util.UUID

data class Price(
    val currency: String,
    val amount: BigDecimal
) {
    init {
        require(currency.length == 3) { "Currency must be a 3-character code" }
        require(amount > BigDecimal.ZERO) { "Amount must be greater than 0" }
    }
}

data class Location(
    val country: String,
    val municipality: String,
    val geohash: String
) {
    init {
        require(country.length == 2) { "Country must be a valid ISO 3166 2-character code" }
        require(municipality.isNotBlank()) { "Municipality cannot be blank" }
        require(geohash.length == 7) { "Geohash must be exactly 7 characters" }
    }
}

data class Listing(
    val listingId: UUID = UUID.randomUUID(),
    val name: String,
    val description: String,
    val price: Price,
    val category: String,
    val location: Location
) {
    init {
        require(name.isNotBlank()) { "Name cannot be blank" }
        require(name.length <= 100) { "Name cannot exceed 100 characters" }
        require(description.isNotBlank()) { "Description cannot be blank" }
        require(VALID_CATEGORIES.contains(category)) { "Invalid category: $category" }
    }

    companion object {
        val VALID_CATEGORIES = setOf(
            "Electronics",
            "Fashion",
            "Home & Garden",
            "Motors",
            "Collectibles & Art",
            "Sporting Goods",
            "Toys & Hobbies",
            "Business & Industrial",
            "Music",
            "Health & Beauty",
            "Books",
            "Cameras & Photo",
            "Computers, Tablets & Networking",
            "Cell Phones & Accessories",
            "Video Games & Consoles"
        )
    }
}

data class PaginatedListingsResponse(
    val items: List<Listing>,
    val totalItems: Int,
    val page: Int,
    val pageSize: Int
)
